polkadot-parachains/parachains-common/src/lib.rs

```diff
@@ -109,3 +116,56 @@ pub mod opaque {
 	/// Opaque block identifier type.
 	pub type BlockId = generic::BlockId<Block>;
 }
+
+//TODO: move DenyThenTry to polkadot's xcm module.
+/// Deny executing the xcm message if it matches any of the Deny filter regardless of anything else.
+/// If it passes the Deny, and matches one of the Allow cases then it is let through.
+pub struct DenyThenTry<Deny, Allow>
+where
+	Deny: ShouldExecute,
+	Allow: ShouldExecute,
+{
+	_deny: PhantomData<Deny>,
+	_allow: PhantomData<Allow>,
+}
+
+impl<Deny, Allow> ShouldExecute for DenyThenTry<Deny, Allow>
+where
+	Deny: ShouldExecute,
+	Allow: ShouldExecute,
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		message: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		if Deny::should_execute(origin, message, max_weight, weight_credit).is_ok() {
```

Comment:
The name of this function is `should_execute`, and so when it returns an Ok, I expect it to be able to execute, not be denied or blocked.

Reply:
You want me to create a ShouldNotExecute trait?

Reply:
No, I think `should_execute` should return an `Err` when it judges that the XCM should be denied from execution.